Ignore SWT tracing function in trace builder and update tests.

Pavel-Durov opened the pull request:
## Changes

- Ignore the tracing function `yk_trace_basicblock` in tracer builder.
- Update ignored tests

## Tests

Tested multiple times as:
```shell
YKB_TRACER=hwt cargo test
```
```shell
YKB_TRACER=hwt YKD_NEW_CODEGEN=1 cargo test
```
```shell
YKB_TRACER=swt cargo test
```
```shell
YKB_TRACER=hwt YKD_NEW_CODEGEN=1 cargo test
```

## Notes
I did see `simple_binop.c` test failing one time with a warning: `yk-warning: trace-compilation-aborted: Trace too long.`
But I don't think it's relevant to my changes in the PR.

ltratt commented on ykrt/src/compile/jitc_yk/trace_builder.rs:
```diff
@@ -765,6 +768,9 @@ impl TraceBuilder {
         if inst.is_debug_call(self.aot_mod) {
             return Ok(());
         }
+        if AOT_MOD.func(*callee).name() == "yk_trace_basicblock" {
```
Should this be `__yk_trace_basicblock`?

Pavel-Durov replied:
maybe, but that's how I defined it before:
https://github.com/ykjit/ykllvm/blob/main/llvm/include/llvm/Transforms/Yk/BasicBlockTracer.h#L7

Pavel-Durov replied:
I can rename it if the naming convention is to start with `__`

ltratt replied:
Yeah we use `__` for such things, but you can do the renaming in a separate PR after this one.

Pavel-Durov replied:
sounds good.

ptersilie commented on ykrt/src/compile/jitc_yk/trace_builder.rs:
```diff
@@ -765,6 +765,9 @@ impl TraceBuilder {
         if inst.is_debug_call(self.aot_mod) {
             return Ok(());
         }
+        if AOT_MOD.func(*callee).name() == "yk_trace_basicblock"{
```
A small comment why we need to do that would be helpful for our future selves, I believe.

Pavel-Durov replied:
Added :point_right: 12e047f93d62f031a865b53816e29df0bc0dc06d

Pavel-Durov commented:
> There's a couple of ignored tests that just say "crashes with SWT". Are these all the same crashes or are they different? What's the plan for these tests? Do we want to fix them here, or will that be a follow up PR?

The purpose of this PR was to update the state of failing tests with SWT.
The next steps will be to triage the failures and come up with a plan to fix them.


The "crashes with SWT" comment was just copied from another source, like this [one](https://github.com/ykjit/yk/blob/master/tests/c/truncate.c#L5). I can remove the `FIXME` comments or add them to all of the tests that are ignored for SWT. Any preference?

Pavel-Durov commented on tests/c/doubleinline.c:
```diff
@@ -1,4 +1,5 @@
-// ignore-if: test $YK_JIT_COMPILER != "yk" -o "$YKB_TRACER" = "swt"
+// ## FIXME: crashes under SWT
```
Actually, this one is actually not terminating at all; it runs for more than a minute. 
Maybe the copied-over comment is misleading here... Remove it?

ltratt replied:
Do we know what's going on that causes it to hang? I suspect a hang in a compiler thread. Does synchronised compilation show anything?

Pavel-Durov replied:
Haven't looked into each test yet,  but I do intend to do it in the following PR. This test hangs with `YKD_SERIALISE_COMPILATION=1`. 

It looks like the control flow is messed up and it just goes into an infinite loop, I suspect that other tests will have the same issue.

The test asserts on:
```
//     3
//     foo-if
//     bar
//     yk-jit-event: enter-jit-code
//     2
//     foo-if
//     bar
//     1
//     yk-jit-event: deoptimise
//     foo-else
//     bar
//     0
//     exit
```

but it prints:
```
...
3
foo-if
bar
yk-jit-event: enter-jit-code
2
foo-if
2
foo-if
2 <------------- INFINITY
```

Pavel-Durov replied:
Attaching output of:
```shell
YKD_SERIALISE_COMPILATION=1 YKB_TRACER=swt /home/pd/.cargo/bin/cargo test ::doubleinline.c  -- --nocapture |& tee out.txt
```
[out.txt](https://github.com/user-attachments/files/16514834/out.txt)

ltratt replied:
If you `gdb` it, where is it stuck?

Pavel-Durov replied:
Maybe we should merge this PR first so that we make sure we don't have regression?
And then I can investigate why tests are failing.

ltratt commented:
@ptersilie Over to you.

ptersilie commented on ykrt/src/compile/jitc_yk/trace_builder.rs:
```diff
@@ -697,6 +697,10 @@ impl TraceBuilder {
+        // Ignore software tracer calls. Software tracer inserts
+        // `yk_trace_basicblock` insturction calls into the begining of
+        // every basic block. These calls can be ignored as they only
+        // used to collect runtime information for the tracer itself.
```
There's a few typos here: s/insturction/instruction, s/begining/beginning, s/they only/they are only. There's no need to push a new commit. Just fix them during squashing.

Pavel-Durov replied:
oops.
fixed :point_right: 541e8ade5bc8742c9e06547afef10d99bdde8e41

ptersilie commented:
Please squash.

Pavel-Durov commented:
squashed :point_right: e0a17f0e37623638f3d88f0b1f92048d5cec110d

Pavel-Durov commented:
Actually, more tests are passing now :eyes: 
:point_right: b6b7f46227b45a75041f7eeeea4c96c63ee40daa

Pavel-Durov commented:
> There's a couple of ignored tests that just say "crashes with SWT". 

Removed in :point_right: 34c1ace6fdb2db4d42a066a000b4149cc3cb6e80
As its easy to find tests that are failing under SWT based on `ignore-if`

ltratt commented:
Please squash.

Pavel-Durov commented:
Squashed :point_right: bf2b45e9a1c2685a3db7b781a6f0155c0e8f652b

Pavel-Durov committed:
Ignore yk_trace_basicblock in trace_builder.

Software tracer inserts tracing instruction calls into the beginning of
every basic block. These calls can be ignored as they are only used to
collect runtime information for the tracer itself.

Pavel-Durov commented:
squashed again :point_right: 367f780b0b01eabd3a810df1a524d90f8562f194
Last squash we had 12 tests ignored for swt now we have 26

Pavel-Durov commented on .buildbot.sh:
```diff
@@ -2,8 +2,7 @@
 
-# FIXME: Reenable swt once jitc_llvm is fully removed.
-TRACERS="hwt"
+TRACERS="hwt swt"
```
I added swt back :)
